### ISSUE TITLE
[GlobalDynamicProperties] Extending the limits of outbound Token Transfers 

### ISSUE THREAD

**Issue by mattsmithies** [hashgraph/hedera-services#1063]
**Summary**

_Consider this a low priority in the short term._

With Unibar we have been exploring introducing a concept of one-sided Liquidity Pools (LP) on top of account creation and HTS.

In particular, we've designed a fair mechanism for distributing and claiming rewards to each LP participant.

In brief, this is the current pool flow:

1. A pool (new account) is created for a given HTS token.
2. The token can be deposited into a pool from any Unibar account holder.
3. For every deposit from a unique holder a new liquidity reward pool is created and mints an LP receipt token which is sent back to the deposited holder.
4. In addition a holder can explicitly choose which assets (think HTS wrapped BTC, ETH, DAI etc) they want for their tokens
5. These "LP Receipt" can be traded back into the reward pool that minted them, to receive the initial token stake and swapped asset gains.
6. A buyer can swap their asset for a given "token".
7. After the buyer has received a "token" the assets are distributed to all "liquidity reward pools" that ask for the given asset based on their ownership of a given swap scenario.

The **TransferTransaction** provides a mechanism to execute this action in a single transaction, to multiple accounts.

### A Scenario

There is a pool for a **UNIBAR** token.
- Lina deposits 100 **UNIBAR** and wants **htsBTC**
- Gehrig deposits 20 **UNIBAR** and wants **htsETH**
- Cooper deposits 80 **UNIBAR** and wants **htsETH**
- Brady wants to spend **htsETH** for UNIBAR, Brady has access to purchase up to 50% of all available **UNIBAR**
- Upon purchase, the **htsETH** are distributed as an 80/20 split to the reward pools that Cooper and Gehrig have the LP receipts respectively as a single purchase.
- (Beyond the current scope) After which Lina could now own a higher stake of the total **UNIBAR** pool and be notified of the sale, as there is a higher demand for purchases via **htsETH** and be offered to add that token but connected to a higher price.

For a small number of participants, this is fine but beyond 8 we have to currently chunk the TokenTransactions.

**Possible resolution**

It seems that the **[ledger.tokenTransfers.maxLen](https://github.com/hashgraph/hedera-services/blob/c4fc617295941a6d34cdb4e2e9a97722ef993bb4/hedera-node/src/main/java/com/hedera/services/context/properties/GlobalDynamicProperties.java#L90)** is set as **8**, as expected any more triggers a **TOKEN_TRANSFER_LIST_SIZE_LIMIT_EXCEEDED** error.

Is there any internal reasons or implications for raising this higher, perhaps initially to 64 and then potentially to 1024/2048 and beyond?

If we look at exchanges like Uniswap and the holders for a given liquidity pool can reach in the thousands.

**Comment by tinker-michaelj** [comment]
Hi @mattsmithies, thanks for raising this issue! I'll need to re-read your use case a few more times to grok it 😁 , but I can give some color on the on the transfer list limit.

The reason for the current limit is our throttling logic. (That is, the [logic](https://github.com/hashgraph/hedera-services/blob/master/hedera-node/src/main/java/com/hedera/services/throttling/BucketThrottling.java#L81) that decides if a node returns `BUSY` or accepts a transaction.) Suppose you submit a transaction. The throttling logic must estimate how much capacity the network has available, and then compare that to how much capacity your transaction will require. 

Right now we take a pretty rough-grained approach that treats all transactions of the same type in the same way. 

In particular, we assume each operation defined in the  [`HederaFunctionality enum`](https://hashgraph.github.io/hedera-protobufs/#proto.HederaFunctionality) requires a **fixed amount of capacity**. (For example, a `CryptoTransfer` might require only `1` unit of capacity, while the far more expensive `ContractCall` might require `400` units.) 

This assumption means there must be a fairly strict **upper bound on the actual resources used** by each transaction type. If we let a `CryptoTransfer` move currencies between anywhere from 2 to 1024 accounts, this assumption would...break down, haha.

**Comment by mattsmithies** [comment]
Hi @tinker-michaelj thank you for the reply,

Understandably there are current limits are imposed by the throttle, my belief is that this would be the case.  I think may have inadvertently bumped into the throttle more times than I have realized during my experiments.

I'm going to use this as an opportunity to further illustrate the general approach, in the hopes that I can make it clearer and more concise for both this thread and for the hackathon presentation. 

One of my initial goals with Unibar, particularly during the hackathon, is to use HTS as a sandbox and explore a design that could show a vision for a Uniswap scale application but on hedera.

Doing this at the experimental risk of being possibly too resource greedy, trying to dance with the current limits of HTS. There is likely a simpler mathematical process of handling the use case.

In short, the aim of the project is around building DeFi "lego" primitives that can be exposed as an accessible API for any product team. Rather than being a market into itself, use pricing data from Ethereum. 

Attached is an image that shows the concept and flow.

![UNIBAR one-sided liquidity pools](https://user-images.githubusercontent.com/6974163/107323637-872a3780-6a9e-11eb-8a62-f01f5b4ef632.png)

**Comment by tinker-michaelj** [comment]
Hey @mattsmithies, thanks for expanding...I still don't have the full picture, I think. 

From what I know of Uniswap, there are essentially two parties involved: 
1. Liquidity providers, who deposit tokens/ETH into "pair contracts".
2. Traders, who can automatically exchange tokens/ETH with pair contracts for a fee.

Traders benefit from the liquidity; and the providers benefit from sharing the fees.

Does Unibar still have these two basic categories of participants? 

I'm **tinkerm** on the [Hedera Discord](https://discord.com/invite/FFb9YFX)...not super-active, but we could discuss more there if you're interested.

**Comment by netopyr** [comment]
Closed due to inactivity. Please reopen if you think the issue is still relevant.
